### PR TITLE
apps: persist selected catalogs when navigating between app pages

### PR DESCRIPTION
**Pull request by axbarsan**
Closes https://github.com/giantswarm/giantswarm/issues/17361

With this PR, the selected catalogs in the sidebar of the `Apps` page will be persisted between route changes (disclaimer: only if we change routes between the app-specific routes - app list and app details, if we navigate outside of that, the selection will be reset).

### CODE REVIEW

File: src/components/MAPI/apps/AppsProvider.tsx [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  createContext,
+  useCallback,
+  useContext,
+  useMemo,
+  useReducer,
+} from 'react';
+
+export type SelectedCatalogs = Record<string, boolean>;
+
+export interface IAppsContext {
```
Comment (axbarsan) [inline src/components/MAPI/apps/AppsProvider.tsx:13]: Made this a little more generic, so we could add some more common app functionality later

**Comment by marians** [comment]
Seeing some exceptions which are supposedly coming from React Dev tools:

```
Warning: Failed prop type: Invalid prop `appsIsLoading` of type `string` supplied to `AppsList`, expected `boolean`.
Warning: Failed prop type: Invalid prop `isLoading` of type `string` supplied to `Toolbar`, expected `boolean`.
```

**Comment by axbarsan** [comment]
> Seeing some exceptions which are supposedly coming from React Dev tools:
> 
> ```
> Warning: Failed prop type: Invalid prop `appsIsLoading` of type `string` supplied to `AppsList`, expected `boolean`.
> Warning: Failed prop type: Invalid prop `isLoading` of type `string` supplied to `Toolbar`, expected `boolean`.
> ```

I already answered that here https://github.com/giantswarm/happa/pull/2441#issuecomment-845823182

**Comment by marians** [comment]
Regarding persisting the list page state

1. The search term should be persisted, too. When I search for "calico" and then navigate into the calico app, then hit Back, I should get back to the search result for "calico".
2. The list sorting selection should be persisted, too.

**Comment by axbarsan** [comment]
> Regarding persisting the list page state
> 
> 1. The search term should be persisted, too. When I search for "calico" and then navigate into the calico app, then hit Back, I should get back to the search result for "calico".
> 2. The list sorting selection should be persisted, too.

Please take a look 👍🏻

**Review by marians — APPROVED** [review]
Much appreciated!

**Pull request merged**